File: Algorithm/Double Pointer/0680 Valid Palindrome II.cpp
class Solution {
public:
  bool validPalindrome(string s) {
    int l = 0, r = s.length()-1; 
    
    while (l < r){
      // when there is a difference
      if (s[l] != s[r]){
        return isPalindrome(s, l+1, r) || isPalindrome(s, l, r-1); 
      }
      else {
        l++; 
        r--; 
      }
    }
    return true; 
  }
    
private: 
  bool isPalindrome(const string& s, int l, int r){
    while (l < r){
      if (s[l++] != s[r--]) return false; 
    }
    return true; 
  }
};

File: Algorithm/Double Pointer/0680 Valid Palindrome II.py
class Solution:
  def validPalindrome(self, s: str) -> bool:
    # initialize a left and right pointer
    l = 0
    r = len(s) - 1
    # loop through the string
    while l < r:
      if s[l] == s[r]:
        l += 1
        r -= 1
      else: 
        # s[l:r][::-1] reverts the string
        # Checking deleting the right char or deleting the left char with return a palindrome
        return s[l:r] == s[l:r][::-1] or s[l+1:r+1] == s[l+1:r+1][::-1]
    return True
                